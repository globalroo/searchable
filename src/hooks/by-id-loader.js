import { useState, useEffect } from "react";

// Centralise byId loading / state management
export const useByIdLoader = ({ id, fetcher }) => {
	const [response, setResponse] = useState({});
	const [error, setError] = useState(false);
	const [loading, setLoading] = useState(true);

	useEffect(() => {
		setError(() => false);
		setLoading(() => true);
		fetcher(id).then(response => {
			if (!response.response.ok) {
				setError(() => true);
			}
			setLoading(() => false);
			setResponse(() => response);
		});
	}, [id]);

	return { loading, error, response };
};
