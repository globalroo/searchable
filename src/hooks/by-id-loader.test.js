import React from "react";
import { render } from "react-testing-library";

import { useByIdLoader } from "./by-id-loader";

const mockLoadedData = {
	id: 1,
	data: "This is a mock component to exercise the hook"
};

// Exercise the hook by ensuring a Fake component is populated as expected
export const HookTestingComponent = ({ id, fetcher }) => {
	const [detail] = useByIdLoader({ id, fetcher });
	return <div>{JSON.stringify(detail)}</div>;
};

describe("byIdLoader hook", () => {
	// TODO: Early adopter issue here - Suppressing misleading async error with jest timers and the new hook functionality.
	// Fix to be published "Real soon now" per https://twitter.com/dan_abramov
	// https://github.com/facebook/react/issues/14769
	const originalError = console.error;

	beforeAll(() => {
		console.error = jest.fn();
	});

	afterAll(() => {
		console.error = originalError;
	});

	it("Uses the loading hook correctly", async () => {
		const fetcher = jest.fn(() => Promise.resolve(mockLoadedData));
		const { container } = await render(<HookTestingComponent id={1} fetcher={fetcher} />);
		expect(container.firstChild).toMatchSnapshot();
	});
});
