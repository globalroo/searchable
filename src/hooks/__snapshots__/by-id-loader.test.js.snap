// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`byIdLoader hook Uses the loading hook correctly 1`] = `
<div>
  {"response":{"payload":{"id":1,"data":"This is a mock component to exercise the hook"}}}
</div>
`;
